### PR TITLE
Review scheduler annotations, part 2

### PR DESCRIPTION
**Pull request by crusaderky**
Continues from #6132 

#### Out of scope:
- SchedulerState
- Scheduler
- A bunch of non-None attributes in TaskState that are forced to None in ``TaskState.__init__``
- Change TaskState.state to a Literal

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -874,363 +858,265 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict:
...
+    __slots__ = tuple(__annotations__)  # type: ignore
 
     def __init__(self, key: str, run_spec: object):
         self.key = key
         self._hash = hash(key)
         self.run_spec = run_spec
-        self._state: str = None  # type: ignore
-        self.exception: str = None  # type: ignore
-        self.exception_blame: TaskState = None  # type: ignore
+        self._state = None  # type: ignore
```
Comment (crusaderky) [inline distributed/scheduler.py:1067]: The cleanup of these non-None attributes forced to None here is left to a future PR

File: distributed/scheduler.py [hunk 2]
```diff
@@ -1948,11 +1829,7 @@ def decide_worker(self, ts: TaskState) -> WorkerState:  # -> WorkerState | None
             else:  # dumb but fast in large case
                 ws = wp_vals[self.n_tasks % n_workers]
 
-        if self.validate:
-            assert ws is None or isinstance(ws, WorkerState), (
-                type(ws),
-                ws,
-            )
+        if self.validate and ws is not None:
             assert ws.address in self.workers
```
Comment (crusaderky) [inline distributed/scheduler.py:1833]: !!! The top-level function ``decide_workers`` NEVER returns None in the unit tests !!!

Comment (crusaderky) [reply distributed/scheduler.py:1833]: https://github.com/dask/distributed/issues/6254

**Comment by crusaderky** [comment]
All failures are unrelated

File: distributed/scheduler.py [hunk 3]
```diff
@@ -877,363 +861,265 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict:
...
-            other_ts: TaskState = other
-            return self.key == other_ts.key
-        else:
-            return False
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, TaskState) and self.key == other.key
 
     @property
     def state(self) -> str:
+        """This task's current state.  Valid states include ``released``, ``waiting``,
```
Comment (hendrikmakait) [inline distributed/scheduler.py:1109]: Thanks for this docstring! Without it, it has been extremely hard for me to understand what possible states a `TaskState` could be in.

**Review by hendrikmakait — APPROVED** [review]
Overall LGTM. I think there are a few more annotations we can drop while we're at it, but feel free to ignore these (as well as my nit-picking and meta-comment):
https://github.com/dask/distributed/blob/2e9971d7a6c563e51327926a20b8ec569978330d/distributed/scheduler.py#L1133

https://github.com/dask/distributed/blob/2e9971d7a6c563e51327926a20b8ec569978330d/distributed/scheduler.py#L7194

https://github.com/dask/distributed/blob/2e9971d7a6c563e51327926a20b8ec569978330d/distributed/scheduler.py#L7266

File: distributed/scheduler.py [hunk 4]
```diff
@@ -7678,78 +7542,91 @@ def validate_state(tasks, workers, clients):
...
+        restrictions: dict[str, float],
+        **kwargs,
+    ) -> None:
         self.keys.update(keys)
 
-    def transition(self, key, start, finish, *args, **kwargs):
+    def transition(self, key: str, start: str, finish: str, *args, **kwargs) -> None:
         if finish == "memory" or finish == "erred":
-            ts: TaskState = self.scheduler.tasks.get(key)
+            ts = self.scheduler.tasks.get(key)
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:7629]: Yes, we ubiquitously intend ts as TaskState and ws as WorkerState
"ts" as timestamp should be avoided in distributed.

File: distributed/scheduler.py (hunk 3 above)
Comment (crusaderky) [reply distributed/scheduler.py:1109]: I'm going to replace str with a TaskStateState Literal in a later PR, like it's already done in worker_state_machine.py

File: distributed/scheduler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:7628]: Alright, so it'll just take some getting used to :)

**Pull request merged**